packages/headless/src/features/facets/facet-set/facet-set-analytics-actions-utils.ts

```diff
@@ -21,3 +43,158 @@ export type FacetSelectionChangeMetadata = {
+  return {
+    facetSet: s.facetSet || getFacetSetInitialState(),
+    categoryFacetSet: s.categoryFacetSet || getCategoryFacetSetInitialState(),
+    dateFacetSet: s.dateFacetSet || getDateFacetSetInitialState(),
+    numericFacetSet: s.numericFacetSet || getNumericFacetSetInitialState(),
+    search: s.search || getSearchInitialState(),
+  };
+}
+
+export const buildFacetStateMetadata = (
+  state: SectionNeededForFacetMetadata
+) => {
+  const facetState: FacetStateMetadata[] = [];
+
+  state.search.response.facets.forEach((facetResponse, facetPosition) => {
+    (facetResponse.values as Array<AnyFacetValue>).forEach(
+      (facetValue, facetValuePositition) => {
+        if (facetValue.state === 'selected') {
+          const facetType = getFacetType(state, facetResponse.facetId);
+
+          const facetResponseAnalytics = mapFacetResponseToAnalytics(
+            facetResponse,
+            facetPosition + 1
+          );
+
+          const facetValueAnalytics = mapFacetValueToAnalytics(
+            facetValue,
+            facetValuePositition + 1,
+            facetType
+          );
+
+          const facetDisplayValueAnalytics =
+            facetType === 'hierarchical' || facetType === 'specific'
+              ? mapFacetDisplayValueToAnalytics(
+                  facetValue as CategoryFacetValue | FacetValue
+                )
+              : mapRangeDisplayFacetValueToAnalytics(
+                  facetValue as NumericFacetValue | DateFacetValue
+                );
+
+          facetState.push({
+            ...facetResponseAnalytics,
+            ...facetValueAnalytics,
+            ...facetDisplayValueAnalytics,
+          });
+        }
+      }
+    );
+  });
+
+  return facetState;
+};
+
+const mapFacetValueToAnalytics = (
+  facetValue: AnyFacetValue,
+  valuePosition: number,
+  facetType: AvailableType
+) => {
+  return {
+    state: facetValue.state,
+    valuePosition,
+    facetType,
+  };
+};
+
+const mapRangeDisplayFacetValueToAnalytics = (
+  facetValue: DateFacetValue | NumericFacetValue
+) => {
+  return {
+    displayValue: `${facetValue.start}..${facetValue.end}`,
+    value: `${facetValue.start}..${facetValue.end}`,
+    start: facetValue.start,
+    end: facetValue.end,
+    endInclusive: facetValue.endInclusive,
+  };
+};
+
+const mapFacetDisplayValueToAnalytics = (
+  facetValue: CategoryFacetValue | FacetValue
+) => {
+  return {
+    displayValue: facetValue.value,
+    value: facetValue.value,
+  };
+};
+
+const mapFacetResponseToAnalytics = (
+  response: AnyFacetResponse,
+  facetPosition: number
+) => {
+  return {
+    title: response.facetId,
+    field: response.field,
+    id: response.facetId,
+    facetPosition,
+  };
+};
+
+const getFacetType = (
+  state: SearchSection &
+    FacetSection &
+    CategoryFacetSection &
+    DateFacetSection &
+    NumericFacetSection,
+  id: string
+): AvailableType => {
+  const facetResponse = facetResponseSelector(state, id);
```

Comment:
Is it necessary to go through the response to determine the type? Given a facet id, we can check the sets directly (this is also what the selectors are doing). The logic to retrieve a facet by id is also being used in `buildFacetBaseMetadata` above, so could be extracted and reused. e.g.

```
const getFacet = (id) => {
    return state.facetSet[facetId] ||
    state.categoryFacetSet[facetId] ||
    state.dateFacetSet[facetId] ||
    state.numericFacetSet[facetId];
}

const getFacetType = () => {
    const facet = getFacet(id)
    return facet ? facet.type : 'specific';
}
```